go/merkletree2/tree.go

```diff
@@ -1,38 +1,405 @@
 package merkletree2
 
-import "sync"
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/net/context"
+)
 
 // Tree is the MerkleTree class; it needs an engine and a configuration
 // to run
 type Tree struct {
 	sync.RWMutex
-	//eng StorageEngine not yet implemented/defined
-	cfg Config
+	TreeConfig
+
+	eng StorageEngine
+	log logger.Logger
 }
 
-// ValueConstructor is an interface for constructing values, so that typed
-// values can be pulled out of the Merkle Tree. We are of course assuming
-// that there is only one type of Value at the leaves, which makes sense.
-type ValueConstructor interface {
-	// Construct a new template empty value for the leaf, so that the
-	// Unmarshalling routine has the correct type template.
-	Construct() interface{}
+// NewTree makes a new tree
+func NewTree(c TreeConfig, e StorageEngine, l logger.Logger) *Tree {
+	return &Tree{TreeConfig: c, eng: e, log: l}
 }
 
 // Hash is a byte-array, used to represent a full collision-resistant hash.
 type Hash []byte
 
+// Equal compares two hashes byte by byte
+func (h Hash) Equal(h2 Hash) bool {
+	return bytes.Equal(h, h2)
+}
+
 // Key is a byte-array, and it is the type of the keys in the KeyValuePairs that
 // the tree can store.
 type Key []byte
 
+// Equal compares two keys byte by byte
+func (k Key) Equal(k2 Key) bool {
+	return bytes.Equal(k, k2)
+}
+
+// Cmp compares two keys lexicographically as byte slices
+func (k Key) Cmp(k2 Key) int {
+	return bytes.Compare(k, k2)
+}
+
 // Seqno is an integer used to differentiate different versions of a merkle tree.
-type Seqno uint
+type Seqno int
```

Comment:
we should pick a type and use it throughout...I don't have too strong of an opinion, maybe uint64?

Reply:
Let's do int64. The reason for the change is that if we do arithmetic over this stuff it is easy to get silent underflows.  `len([]byte)` returns an int rather than uint for this reason.